### PR TITLE
Add: darkmode support + other GUI related improvements

### LINKED ISSUE #796: Implement dark mode in the UI
The Swing theme we are using has a dark mode: https://www.formdev.com/flatlaf/

We should enable a dark mode toggle somewhere in the UI, taking care of refreshing the look and feel after it is switched: https://docs.oracle.com/javase/tutorial/uiswing/lookandfeel/plaf.html#dynamic

### PR DESCRIPTION
**Pull request by smallg0at**
image

image

This adds darkmode support and closes #796, and due to how I worked on this #800 might need to be merged first

Feature introduced:
- A toggle to enable dark mode, which take effect on restart
- The restart prompt for settings have been renamed
- Redraw all window icons 

Problems still existing:
- ~~Default highlight colors are way too bright, lacking any contrast with the text wrapped by it.~~ The current solution addresses part of the problem, but the user interface design still needs further discussion.
- Loading up a file after theme change in settings will cause the theme of Cycles and Pipeline window to misalign with the main window
- ~~Icons does not match the design of this new UI~~
- ~~I/O Window is... image~~ fixed

### CODE REVIEW

**Comment by smallg0at** [comment]
image

This seems optimal, though user preference is somewhat ignored as it uses darker() three times

**Comment by smallg0at** [comment]
image

Icons have been redesigned from the ground up for better presentation & to adapt to dark mode

**Comment by lupino3** [comment]
Thanks for your work on this PR! I want to release version 1.3, because we radically changed the semantics of `DMULU` and there are other fixes, but I'll wait to see if we can get dark mode in as well. Thanks!

**Comment by lupino3** [comment]
> image
> 
> This seems optimal, though user preference is somewhat ignored as it uses darker() three times

I would like to test this a bit when I get some time.. It might be sufficient to say that the color chosen by the user is darkened in dark mode. Or maybe we need the user to get other colors for the dark mode. Or, finally, maybe we should provide defaults that work for both modes, and allow the user to change them if they really want. (which would mean we need to use different internal config names or we'll pick what the user had chosen in the past)

**Comment by smallg0at** [comment]
> > image
> > This seems optimal, though user preference is somewhat ignored as it uses darker() three times.
> 
> I would like to test this a bit when I get some time. It might be sufficient to say that the color chosen by the user is darkened in dark mode. Or maybe we need the user to get other colors for the dark mode. Or, finally, maybe we should provide defaults that work for both modes and allow the user to change them if they really want. (which would mean we need to use different internal config names, or we'll pick what the user had chosen in the past)

- Allow user to get other colors exclusive for the dark mode is probably not a great idea, as edumips64 is not a software with a complete theme system that covers the entire UI (like VSCode). This would only bloat the setting page where the user will be needed to set the color twice.
- Dimming the color with `darker()` is possible but this specific method lacks detailed customization (no args). Two phases are a little too bright and three is a little to dark to be noticed.
- I'll try to find if one-color-fits-both solution is possible.

**Comment by smallg0at** [comment]
image

Webapp is receiving an update as well

**Comment by lupino3** [comment]
> image
> 
> Webapp is receiving an update as well

Woah... THANK YOU!

**Comment by lupino3** [comment]
image

the new accordion-based interface makes SO MUCH MORE SENSE than what I quickly threw together.

Thank you, you are amazing!

**Comment by lupino3** [comment]
Small note: I think the bottom of the web UI is not visible anymore, the widgets seem to extend outside of the web browser view.

**Comment by smallg0at** [comment]
# Finalizing Webapp

The development related to this webapp in the PR has concluded.

## Changes

- Redesigned UI placements, allowing this thing to at least look like an IDE
- Adds dark mode, which will always match system / browser preference on this. Not manually tunable.
- Now user is allowed to manually specify how many step to run in Multi-Step
- Version code gets moved to the Help dialog

**Comment by smallg0at** [comment]
> Small note: I think the bottom of the web UI is not visible anymore, the widgets seem to extend outside of the web browser view.

I cannot replicate this issue so far, I might need a screenshot of it happening

**Comment by lupino3** [comment]
> # Finalizing Webapp
> The development related to this webapp in the PR has concluded.
> 
> ## Changes
> * Redesigned UI placements, allowing this thing to at least look like an IDE
> * Adds dark mode, which will always match system / browser preference on this. Not manually tunable.
> * Now user is allowed to manually specify how many step to run in Multi-Step
> * Version code gets moved to the Help dialog

Since this PR has both webapp and (incomplete) desktop changes, would you like to split out the webapp changes in a separate PR so I can merge it sooner?

**Comment by lupino3** [comment]
image

This is how it looks like if I go full-screen and then resize down.

**Comment by smallg0at** [comment]
Fixed. Turns out flexGrow panics out and took an entire screen

**Comment by smallg0at** [comment]
## Known Issue

- Due to the nature of EduMips64's logo, its contrasts are terrible in dark mode

**Comment by smallg0at** [comment]
> image
> 
> This is how it looks like if I go full-screen and then resize down.

Overlooked the error counter component, going to look into it

**Comment by lupino3** [comment]
> ## Known Issue
> * Due to the nature of EduMips64's logo, its contrasts are terrible in dark mode

I was about to comment exactly about this :)

image

It's fine, just open an issue about it and we'll fix it later. This looks amazing, thank you!

**Comment by smallg0at** [comment]
> > # Finalizing Webapp
> > The development related to this webapp in the PR has concluded.
> > ## Changes
> > 
> > * Redesigned UI placements, allowing this thing to at least look like an IDE
> > * Adds dark mode, which will always match system / browser preference on this. Not manually tunable.
> > * Now user is allowed to manually specify how many step to run in Multi-Step
> > * Version code gets moved to the Help dialog
> 
> Since this PR has both webapp and (incomplete) desktop changes, would you like to split out the webapp changes in a separate PR so I can merge it sooner?

Working on it, hope it won't blow up

**Comment by smallg0at** [comment]
Webapp update is now #810

**Comment by smallg0at** [comment]
@lupino3 Conclusion: one-color-for-all is not possible, will probably use current approach.

By the way, how can I reset those settings?

**Comment by lupino3** [comment]
It depends on the OS, I am not sure how to do it everywhere. See for example https://apple.stackexchange.com/questions/129489/how-do-i-delete-java-application-preferences

**Comment by smallg0at** [comment]
@lupino3 This PR is ready to be merged

**Comment by lupino3** [comment]
Some initial feedback based on a quick test
image

First of all, this looks great, thank you! Love the new icons and overall the dark theme looks easier on the eyes.

1. The gray of the Cycles and Pipeline windows is different from the one in the other windows. I think it would make sense to centralize color management in a single class rather than instantiating colors throughout the code.
1. The white used in the GUICycles window is too bright, I think we need to make it darker in order to be easier on the eyes.
1. Same for the text in the Cycles and Pipeline window, it's far too bright, while the text in other windows is light grey so it's perfect.
1. Part of the Statistics window is still white, not sure why.
1. Shouldn't the colors in the Cycles/Pipeline window also get darker, like the ones in the other windows?

File: .github/workflows/ci.yml
Comment (lupino3) [inline .github/workflows/ci.yml:1]: Please revert changes on this file.

File: src/main/java/org/edumips64/utils/CurrentLocale.java [hunk 1]
```diff
@@ -440,6 +441,7 @@ public static void setConfig(ConfigStore config) {
     it.put("Config.SYNCEXC-TERMINATE.tip", "Ferma la simulazione al verificarsi di eccezioni di tipo Divisione per zero ed Integer overflow");
     it.put("Config.FONTSIZE", "Dimensione font");
     it.put("Config.FONTSIZE.tip", "Dimensione del font");
+    it.put("Config.DARKTHEME", "Dark theme (take effect after restart)");
```
Comment (lupino3) [inline src/main/java/org/edumips64/utils/CurrentLocale.java:444]:
```suggestion
    it.put("Config.DARKTHEME", "Tema scuro (applicato al prossimo avvio)");
```

**Review by lupino3 — CHANGES_REQUESTED** [review]
Please apply those 2 small changes. Thanks!

**Comment by lupino3** [comment]
> Some initial feedback based on a quick test image
> 
> First of all, this looks great, thank you! Love the new icons and overall the dark theme looks easier on the eyes.
> 
> 1. The gray of the Cycles and Pipeline windows is different from the one in the other windows. I think it would make sense to centralize color management in a single class rather than instantiating colors throughout the code.
> 2. The white used in the GUICycles window is too bright, I think we need to make it darker in order to be easier on the eyes.
> 3. Same for the text in the Cycles and Pipeline window, it's far too bright, while the text in other windows is light grey so it's perfect.
> 4. Part of the Statistics window is still white, not sure why.
> 5. Shouldn't the colors in the Cycles/Pipeline window also get darker, like the ones in the other windows?

Let me know if you have time to work on those changes, or if you'd rather merge the PR as-is (addressing the feedback I left in the review) and we can work on the remaining changes in another PR. Thanks!

**Comment by smallg0at** [comment]
I have time working on this, so I'll resolve them in this PR.

> 1. The gray of the Cycles and Pipeline windows is different from the one in the other windows. I think it would make sense to centralize color management in a single class rather than instantiating colors throughout the code.

- [ ] Fix the color of the two panels. They currently use Color.Darkgrey and therefore can be customized to match theme.

> 2. The white used in the GUICycles window is too bright, I think we need to make it darker in order to be easier on the eyes.

- [ ] investigation planned

> 3. Same for the text in the Cycles and Pipeline window, it's far too bright, while the text in other windows is light grey so it's perfect.

- [ ] fixable, planned

> 4. Part of the Statistics window is still white, not sure why.

This was never experienced on my side. I assume that changing window size will fix this.

> 5. Shouldn't the colors in the Cycles/Pipeline window also get darker, like the ones in the other windows?

- [ ] investigation planned

**Comment by lupino3** [comment]
Thanks for the improvements! The only thing missing is the Statistics window being white for half:

image

The behavior is clear when you make the window larger

image

**Comment by smallg0at** [comment]
> You may need to fix this here:
> 
> https://github.com/smallg0at/edumips64/blob/5254f01d6e4c976d41cda0ef92672d60a603ec28/src/main/java/org/edumips64/ui/swing/GUIStatistics.java#L70
> 
> Should be full-width, I think.

https://github.com/smallg0at/edumips64/blob/5254f01d6e4c976d41cda0ef92672d60a603ec28/src/main/java/org/edumips64/ui/swing/GUIStatistics.java#L68

It's this one's fault.

**Comment by smallg0at** [comment]
image

Now the text has proper AA.

**Comment by smallg0at** [comment]
The two ErrorDialog images uses Microsoft's COLRv1 Emoji, not sure if it will bring any licensing issue

File: src/main/resources/images/error.png
Comment (lupino3) [inline src/main/resources/images/error.png:1]: Is there a reason why you are changing these 2 icons? We created them on purpose for EduMIPS64 and I'd like to keep them :)

Comment (smallg0at) [reply src/main/resources/images/error.png:1]: The old icon does not suit the style of a modern UI, while looking messy when AA is applied. I'll bump up the resolution of the two images anyway, so I replaced them entirely.

So, what's the purpose, or what meanings are there in those icons?

Comment (smallg0at) [reply src/main/resources/images/error.png:1]: btw, I saw a few unused images (errornn, error33, error4), I'm pretty curious what they are for.

Comment (lupino3) [reply src/main/resources/images/error.png:1]: it is mostly a historical artifact of the team that created EduMIPS64 :)
it's nothing important, but it's part of the "character" of the simulator. They are meant to be a funny way of delivering error messages to the user.

I'd love to maintain them, if possible.
The other, unused icons are other possible designs for the error/warning icons that have not been chosen.

Comment (lupino3) [reply src/main/resources/images/error.png:1]: one potential issue I see is that the error icon does contain a "death" reference in a Christian setting (the cross). I am not sure if that can be considered offensive to folks who don't appreciate this kind of references in a tool they are likely mandated to use for their studies.

Comment (smallg0at) [reply src/main/resources/images/error.png:1]: Got it. Here's my decision:
- Rollback changes.
- Try to make the image less blurry / pixelated

**Comment by lupino3** [comment]
Do you want me to merge the changes as they are, or are you working on improving the existing icons? Thanks!

**Comment by smallg0at** [comment]
> Do you want me to merge the changes as they are, or are you working on improving the existing icons? Thanks!

Currently working on the icon. This is a work-in-progress image with AI Upscaling.

warning (1)

**Comment by smallg0at** [comment]
Upscaled images, though ImageIcon is handled weirdly and decided to go 650\*500 -> **130\*100 -> 260\*200** (onscreen at 200% scale, which is made WITHOUT any AA mesaures)
warning-hires
error-hires

**Pull request merged**